bouncer/src/repo/manipulator/modeloptimizer/repo_optimizer_multipart.cpp

```diff
@@ -399,75 +917,234 @@ bool MultipartOptimizer::isTransparent(
 	return isTransparent;
 }
 
+void MultipartOptimizer::clusterMeshNodes(
+	const std::vector<repo::core::model::MeshNode>& meshes,
+	std::vector<std::vector<repo::core::model::MeshNode>>& clusters
+)
+{
+	// Takes one set of MeshNodes and groups them into N sets of MeshNodes, 
+	// where each set will form a Supermesh.
+
+	repoInfo << "Clustering " << meshes.size() << " meshes...";
+	auto start = boost::chrono::high_resolution_clock::now();
+
+	// Start by creating a metric for each mesh which will be used to group
+	// nodes by their efficiency
+
+	struct mesh {
+		repo::core::model::MeshNode node;
+		float efficiency;
+	};
+
+	std::vector<mesh> metrics(meshes.size());
+	for (size_t i = 0; i < meshes.size(); i++)
+	{
+		auto& mesh = metrics[i];
+		mesh.node = meshes[i];
+		auto bounds = mesh.node.getBoundingBox();
+		auto numVertices = mesh.node.getNumVertices();
+		auto width = bounds[1].x - bounds[0].x;
+		auto height = bounds[1].y - bounds[0].y;
+		auto length = bounds[1].z - bounds[0].z;
+		auto metric = (float)numVertices / (width + height + length);
+		mesh.efficiency = metric;
+	}
+
+	// Next order the outer nodes by their efficiency
+
+	std::sort(metrics.begin(), metrics.end(), [](mesh a, mesh b) {
```

Comment:
Set is tricky on first look because it doesn't implement [], but actually rewriting the loops to use iterators makes them neater anyway for the same reasons discussed above!

Reply:
OK, turns out this doesn't work because std::set uses reflexive comparison to figure out if two keys are identical, leading to it dropping meshMetric instances. We could modify the comparator but then we have to detect when efficiencies are equal and add something like the pointer value, which will cause similar efficiencies to jump all other the place. I think the most straightforward solution is to put the vector back unfortunately. (The iterators below are still nice though)